content/4_Gold/LIS.mdx

```diff
@@ -338,6 +338,62 @@ class PCB{
 
 </LanguageSection>
 
+## Extension - LCS on Permutations
+
+Given two permutations of length $N$ ($1\le N \le 10^5$), find the longest common subsequence.
+
+### $\mathcal O(N^2)$
+
+See [here](/gold/paths-grids#solution---longest-common-subsequence).
+
+### $\mathcal O(N\log N)$
+
+In the $\mathcal O(N^2)$ solution, we increase the DP arrray only if $a[i] = b[j]$, but since both arrays are permutations of length $N$, for each $a[i]$, there must be only one matching element in $b$.  
+
+Thus we can reorder the elements in $b$ such that $b[i]$ becomes $\texttt{pos}[b[i]]$ where $\texttt{pos}[i]$ represents the index of $i$ in $a$. After reordering $b$, our answer would be the LIS of $b$. 
```

Comment:
More intuitive if we create array $c[i]$ where $c[i] = a^-1[b[i]]$

Additionally, possibly mention that the final LIS $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the LCS $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$

Reply:
in other words, every increasing subsequence $x_{1 \dots k}$ in $c$ corresponds to a common subsequence between $a$ and $b$

Reply:
> More intuitive if we create array $c[i]$ where $c[i] = a^-1[b[i]]$
> 
> Additionally, possibly mention that the final LIS $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the LCS $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$

what does $c[i] = a^-1[b[i]]$ mean?

Reply:
a^{-1}[i] = pos[i] :/

Reply:
Oops yea that's what I meant. Basically store the "reordered" b array into a new array because it means something different than just reordering it.